File: src/loan_demo.py
"""
Loan Durable Workflow — Frontend Presentation
==============================================

Durable loan approval workflow designed for live demos with a React frontend.

Features:
  - Writes progress to DynamoDB at each step (frontend polls for updates)
  - Hardcoded scenarios based on SIN last 4 digits for predictable outcomes
  - Deliberate time.sleep() in each step to visualize progress
  - External fraud check via callback (separate Lambda sends callback to resume)
  - Manager approval callback for loans >= $100K

Hardcoded scenarios:
  - SIN ending 1111 (Alice): Always approved
  - SIN ending 2222 (Bob):   Always denied (credit score too low)
  - SIN ending 3333 (Charlie): Approved if loan_amount <= $25,000
"""

import hashlib
import json
import os
import time
from datetime import datetime, timezone
from decimal import Decimal

import boto3
from aws_lambda_powertools import Logger

from aws_durable_execution_sdk_python import (
    DurableContext,
    StepContext,
    durable_execution,
    durable_step,
)
from aws_durable_execution_sdk_python.config import (
    Duration,
    WaitForCallbackConfig,
)


logger = Logger()


# ─────────────────────────────────────────────────
# DynamoDB Progress Logging
# ─────────────────────────────────────────────────

def get_progress_table():
    dynamodb = boto3.resource("dynamodb")
    return dynamodb.Table(os.environ["PROGRESS_TABLE"])


def log_progress(table, application_id, step, message, status, level="info", result=None):
    """Append a log entry and update status in DynamoDB."""
    timestamp = datetime.now(timezone.utc).isoformat()
    log_entry = {
        "timestamp": timestamp,
        "step": step,
        "message": message,
        "level": level,
    }

    update_expr = (
        "SET #logs = list_append(if_not_exists(#logs, :empty_list), :new_log), "
        "current_step = :step, #status = :status, updated_at = :ts"
    )
    expr_values = {
        ":new_log": [log_entry],
        ":empty_list": [],
        ":step": step,
        ":status": status,
        ":ts": timestamp,
    }
    expr_names = {
        "#logs": "logs",
        "#status": "status",
    }

    if result is not None:
        update_expr += ", #result = :result"
        expr_values[":result"] = _convert_floats(result)
        expr_names["#result"] = "result"

    table.update_item(
        Key={"application_id": application_id},
        UpdateExpression=update_expr,
        ExpressionAttributeValues=expr_values,
        ExpressionAttributeNames=expr_names,
    )


def _convert_floats(obj):
    """Recursively convert float values to Decimal for DynamoDB."""
    if isinstance(obj, float):
        return Decimal(str(obj))
    if isinstance(obj, dict):
        return {k: _convert_floats(v) for k, v in obj.items()}
    if isinstance(obj, list):
        return [_convert_floats(i) for i in obj]
    return obj


# ─────────────────────────────────────────────────
# Hardcoded Scenario Logic
# ─────────────────────────────────────────────────

def get_scenario_decision(ssn_last4, loan_amount):
    """Return the predetermined decision based on SIN last 4 digits."""
    if ssn_last4 == "1111":
        return "approved"
    elif ssn_last4 == "2222":
        return "denied"
    elif ssn_last4 == "3333":
        return "approved" if loan_amount <= 25000 else "denied"
    # Default: approve for unknown SINs
    return "approved"


# ─────────────────────────────────────────────────
# Durable Steps
# ─────────────────────────────────────────────────

@durable_step
def validate_application(step_context: StepContext, application: dict) -> dict:
    """Validate the loan application fields."""
    logger.info(
        "Validating application",
        extra={"application_id": application.get("application_id")},
    )
    time.sleep(2)

    required = [
        "application_id", "applicant_name", "ssn_last4",
        "annual_income", "loan_amount", "loan_purpose",
    ]
    missing = [f for f in required if f not in application]
    if missing:
        raise ValueError(f"Missing required fields: {missing}")

    income = application["annual_income"]
    loan_amount = application["loan_amount"]

    if loan_amount <= 0:
        raise ValueError("Loan amount must be positive")
    if income <= 0:
        raise ValueError("Annual income must be positive")

    dti_estimate = (loan_amount * 0.05) / (income / 12)

    return {
        "application_id": application["application_id"],
        "applicant_name": application["applicant_name"],
        "ssn_last4": application["ssn_last4"],
        "annual_income": income,
        "loan_amount": loan_amount,
        "loan_purpose": application["loan_purpose"],
        "estimated_dti": round(dti_estimate, 2),
        "status": "validated",
        "validated_at": datetime.now(timezone.utc).isoformat(),
    }


@durable_step
def pull_credit_report(step_context: StepContext, bureau: str, ssn_last4: str) -> dict:
    """Pull a credit report from one bureau."""
    logger.info(f"Pulling credit report from {bureau}")
    time.sleep(3)

    import random
    seed = int(hashlib.md5(f"{bureau}-{ssn_last4}".encode()).hexdigest()[:8], 16)
    rng = random.Random(seed)
    score = rng.randint(580, 820)

    return {
        "bureau": bureau,
        "score": score,
        "report_id": f"{bureau[:3].upper()}-{ssn_last4}-{score}",
        "derogatory_marks": rng.randint(0, 3),
        "open_accounts": rng.randint(2, 15),
        "pulled_at": datetime.now(timezone.utc).isoformat(),
    }


@durable_step
def calculate_risk_score(step_context: StepContext, credit_reports: list, ssn_last4: str, loan_amount: float) -> dict:
    """Aggregate credit reports and apply hardcoded scenario override."""
    logger.info("Calculating risk score from credit reports")
    time.sleep(2)

    scores = [r["score"] for r in credit_reports]
    avg_score = sum(scores) / len(scores)
    total_derogatory = sum(r["derogatory_marks"] for r in credit_reports)

    if avg_score >= 740 and total_derogatory == 0:
        tier, base_rate = "prime", 5.25
    elif avg_score >= 670:
        tier, base_rate = "near-prime", 7.50
    elif avg_score >= 580:
        tier, base_rate = "subprime", 11.00
    else:
        tier, base_rate = "deep-subprime", 15.00

    # Apply hardcoded scenario override
    decision = get_scenario_decision(ssn_last4, loan_amount)

    return {
        "average_score": round(avg_score, 1),
        "min_score": min(scores),
        "max_score": max(scores),
        "total_derogatory_marks": total_derogatory,
        "risk_tier": tier,
        "base_rate": base_rate,
        "decision": decision,
    }


@durable_step
def generate_loan_offer(step_context: StepContext, app: dict, risk: dict) -> dict:
    """Generate the final loan offer with rate and payment terms."""
    logger.info(f"Generating offer for {app['application_id']}")
    time.sleep(2)

    rate = risk["base_rate"]
    loan_amount = app["loan_amount"]
    term_months = 60
    monthly_rate = rate / 100 / 12

    if monthly_rate > 0:
        payment = loan_amount * (
            monthly_rate * (1 + monthly_rate) ** term_months
        ) / ((1 + monthly_rate) ** term_months - 1)
    else:
        payment = loan_amount / term_months

    offer_id = hashlib.sha256(
        f"offer-{app['application_id']}-{rate}".encode()
    ).hexdigest()[:10]

    return {
        "offer_id": f"OFFER-{offer_id.upper()}",
        "application_id": app["application_id"],
        "loan_amount": loan_amount,
        "annual_rate": rate,
        "term_months": term_months,
        "monthly_payment": round(payment, 2),
        "total_interest": round(payment * term_months - loan_amount, 2),
        "status": "offer_generated",
        "generated_at": datetime.now(timezone.utc).isoformat(),
    }


@durable_step
def request_manager_approval(step_context: StepContext, callback_id: str, application_id: str, loan_amount: float) -> dict:
    """Store callback_id in DynamoDB so the frontend can send the approval callback."""
    logger.info(
        f"Manager approval required for ${loan_amount:,.0f} loan "
        f"— callback_id: {callback_id}"
    )

    # Write callback_id to DynamoDB so the API can read it when manager approves
    table = get_progress_table()
    table.update_item(
        Key={"application_id": application_id},
        UpdateExpression="SET callback_id = :cid",
        ExpressionAttributeValues={":cid": callback_id},
    )

    return {
        "application_id": application_id,
        "callback_id": callback_id,
        "status": "pending_manager_approval",
        "requested_at": datetime.now(timezone.utc).isoformat(),
    }


@durable_step
def request_fraud_check(step_context: StepContext, callback_id: str, application_id: str, applicant_name: str) -> dict:
    """Invoke the external fraud check Lambda, passing the callback_id so it can resume us."""
    logger.info(
        f"Requesting external fraud check for {applicant_name} "
        f"— callback_id: {callback_id}"
    )

    lambda_client = boto3.client("lambda")
    fraud_check_function = os.environ["FRAUD_CHECK_FUNCTION"]

    import json
    lambda_client.invoke(
        FunctionName=fraud_check_function,
        InvocationType="Event",
        Payload=json.dumps({
            "callback_id": callback_id,
            "application_id": application_id,
            "applicant_name": applicant_name,
        }),
    )

    return {
        "application_id": application_id,
        "callback_id": callback_id,
        "status": "fraud_check_requested",
        "requested_at": datetime.now(timezone.utc).isoformat(),
    }


@durable_step
def disburse_funds(step_context: StepContext, offer: dict) -> dict:
    """Finalize and disburse the loan — no random failures in demo mode."""
    logger.info(
        f"Disbursing ${offer['loan_amount']:,.0f} for {offer['offer_id']}"
    )
    time.sleep(2)

    return {
        "offer_id": offer["offer_id"],
        "disbursement_ref": f"DSB-{offer['offer_id'][-6:]}",
        "amount_disbursed": offer["loan_amount"],
        "status": "funded",
        "funded_at": datetime.now(timezone.utc).isoformat(),
    }


# ─────────────────────────────────────────────────
# Main Durable Execution Handler
# ─────────────────────────────────────────────────

@durable_execution
def lambda_handler(event: dict, context: DurableContext) -> dict:
    """
    Loan approval workflow with DynamoDB progress logging.
    Invoked asynchronously by the API Lambda.
    """
    table = get_progress_table()
    application_id = event["application_id"]
    logger.append_keys(application_id=application_id)

    # Detect replay: count how many log entries exist per step in DynamoDB.
    # On each invocation, track how many times we've logged each step.
    # If we've already logged that many entries for a step, it's a replay.
    from collections import Counter
    existing = table.get_item(Key={"application_id": application_id}).get("Item", {})
    prior_counts = Counter(entry["step"] for entry in existing.get("logs", []))
    call_counts = Counter()

    def log(step, message, status, level="info", result=None):
        call_counts[step] += 1
        if call_counts[step] <= prior_counts.get(step, 0):
            message = f"[REPLAY] {message}"
            level = "replay"
        log_progress(table, application_id, step, message, status, level, result)

    try:
        # ── Step 1: Validate Application ────────────────────
        log(  "validating", "Validating loan application...", "processing")
        validated = context.step(validate_application(event))

        logger.info(
            f"Application validated: {validated['application_id']} "
            f"— ${validated['loan_amount']:,.0f} ({validated['loan_purpose']})"
        )
        log("validating", "Application validated successfully", "processing")

        # ── Step 2: Parallel Credit Bureau Checks ───────────
        log("credit_check", "Pulling credit reports from 3 bureaus...", "processing")
        bureaus = ["equifax", "transunion", "experian"]

        credit_results = context.parallel([
            lambda ctx, b=b: ctx.step(
                pull_credit_report(b, validated["ssn_last4"])
            )
            for b in bureaus
        ])

        credit_reports = credit_results.get_results()
        scores_str = ", ".join(f"{r['bureau']}={r['score']}" for r in credit_reports)
        logger.info(f"Credit reports pulled — scores: {scores_str}")
        log("credit_check", f"Credit scores received: {scores_str}", "processing")

        # ── Step 3: Risk Assessment ─────────────────────────
        log("risk_assessment", "Calculating risk score...", "processing")
        risk = context.step(calculate_risk_score(
            credit_reports, validated["ssn_last4"], validated["loan_amount"]
        ))

        logger.info(
            f"Risk assessed: tier={risk['risk_tier']}, "
            f"avg={risk['average_score']}, decision={risk['decision']}"
        )
        log(
            "risk_assessment",
            f"Risk tier: {risk['risk_tier']}, avg score: {risk['average_score']}, decision: {risk['decision']}",
            "processing",
        )

        # ── Denied Path ─────────────────────────────────────
        if risk["decision"] == "denied":
            final_result = {
                "application_id": validated["application_id"],
                "applicant_name": validated["applicant_name"],
                "status": "denied",
                "reason": f"Application denied — risk tier: {risk['risk_tier']}, avg credit score: {risk['average_score']}",
                "risk_tier": risk["risk_tier"],
                "average_score": risk["average_score"],
            }
            log(
                "risk_assessment", "Application denied",
                "denied", level="warn", result=final_result,
            )
            return final_result

        # ── Step 4: Manager Approval (if >= $100,000) ────────
        if validated["loan_amount"] >= 100000:
            log(
                "manager_approval",
                f"Manager approval required for loans >= $100,000 (requested: ${validated['loan_amount']:,.0f})",
                "pending_approval",
            )

            def submit_manager_approval(callback_id, _ctx):
                """Store callback_id in DynamoDB so the frontend can send the approval."""
                tbl = get_progress_table()
                tbl.update_item(
                    Key={"application_id": validated["application_id"]},
                    UpdateExpression="SET callback_id = :cid",
                    ExpressionAttributeValues={":cid": callback_id},
                )

            approval_result = context.wait_for_callback(
                submit_manager_approval,
                name="manager-approval",
                config=WaitForCallbackConfig(timeout=Duration.from_minutes(30)),
            )

            if isinstance(approval_result, str):
                approval_result = json.loads(approval_result)
            logger.info(f"Manager approval result: {approval_result}")

            if not approval_result.get("approved"):
                final_result = {
                    "application_id": validated["application_id"],
                    "applicant_name": validated["applicant_name"],
                    "status": "denied",
                    "reason": approval_result.get("reason", "Manager denied the application"),
                }
                log(
                    "manager_approval", "Application denied by manager",
                    "denied", level="warn", result=final_result,
                )
                return final_result

            log("manager_approval", "Manager approved the application", "processing")

        # ── Step 5: External Fraud Check (Callback) ─────────
        # The workflow SUSPENDS here. An external Lambda (FraudCheckFunction)
        # processes the request and calls SendDurableExecutionCallbackSuccess
        # to resume this execution. Zero compute cost while waiting.
        log("fraud_check", "Requesting external fraud check service...", "processing")

        def submit_fraud_check(callback_id, _ctx):
            """Invoke the external fraud check Lambda, passing the callback_id."""
            fraud_lambda = boto3.client("lambda")
            fraud_lambda.invoke(
                FunctionName=os.environ["FRAUD_CHECK_FUNCTION"],
                InvocationType="Event",
                Payload=json.dumps({
                    "callback_id": callback_id,
                    "application_id": validated["application_id"],
                    "applicant_name": validated["applicant_name"],
                }),
            )

        fraud_result = context.wait_for_callback(
            submit_fraud_check,
            name="fraud-check",
            config=WaitForCallbackConfig(timeout=Duration.from_minutes(5)),
        )

        if isinstance(fraud_result, str):
            fraud_result = json.loads(fraud_result)
        logger.info(f"Fraud check result: {fraud_result}")
        log(
            "fraud_check",
            f"Fraud check passed — {fraud_result.get('checked_by', 'external service')}",
            "processing",
        )

        # ── Step 6: Generate Loan Offer ─────────────────────
        log("generating_offer", "Generating loan offer...", "processing")
        offer = context.step(generate_loan_offer(validated, risk))

        logger.info(
            f"Offer generated: {offer['offer_id']} — "
            f"${offer['monthly_payment']}/mo at {offer['annual_rate']}%"
        )
        log(
            "generating_offer",
            f"Offer {offer['offer_id']}: ${offer['monthly_payment']}/mo at {offer['annual_rate']}%",
            "processing",
        )

        # ── Step 6: Disburse Funds ──────────────────────────
        log("disbursing", "Disbursing funds...", "processing")
        disbursement = context.step(disburse_funds(offer))

        logger.info(f"Funds disbursed: {disbursement['disbursement_ref']}")

        final_result = {
            "application_id": validated["application_id"],
            "applicant_name": validated["applicant_name"],
            "status": "approved",
            "offer_id": offer["offer_id"],
            "loan_amount": offer["loan_amount"],
            "annual_rate": offer["annual_rate"],
            "monthly_payment": offer["monthly_payment"],
            "term_months": offer["term_months"],
            "disbursement_ref": disbursement["disbursement_ref"],
        }
        log(
            "complete", "Loan approved and funds disbursed!",
            "approved", result=final_result,
        )
        return final_result

    except Exception as error:
        logger.error(f"Loan workflow failed: {error}")
        log(
            "error", f"Workflow error: {str(error)}",
            "failed", level="error",
        )
        raise

File: src/loan_demo.mjs
/**
 * Loan Durable Workflow — Frontend Presentation
 * ==============================================
 *
 * Durable loan approval workflow designed for live demos with a React frontend.
 *
 * Features:
 *   - Writes progress to DynamoDB at each step (frontend polls for updates)
 *   - Hardcoded scenarios based on SIN last 4 digits for predictable outcomes
 *   - Deliberate delays in each step to visualize progress
 *   - External fraud check via callback (separate Lambda sends callback to resume)
 *   - Manager approval callback for loans >= $100K
 *
 * Hardcoded scenarios:
 *   - SIN ending 1111 (Alice): Always approved
 *   - SIN ending 2222 (Bob):   Always denied (credit score too low)
 *   - SIN ending 3333 (Charlie): Approved if loan_amount <= $25,000
 */

import { Logger } from "@aws-lambda-powertools/logger";
import { DynamoDBClient } from "@aws-sdk/client-dynamodb";
import { DynamoDBDocumentClient, GetCommand, UpdateCommand } from "@aws-sdk/lib-dynamodb";
import { LambdaClient, InvokeCommand } from "@aws-sdk/client-lambda";
import { withDurableExecution } from "@aws/durable-execution-sdk-js";
import crypto from "crypto";

const logger = new Logger({ serviceName: "LoanWorkflow" });


// ─────────────────────────────────────────────────
// DynamoDB Progress Logging
// ─────────────────────────────────────────────────

function getDocClient() {
  const client = new DynamoDBClient({});
  return DynamoDBDocumentClient.from(client);
}

async function logProgress(docClient, applicationId, step, message, status, level = "info", result = null) {
  const timestamp = new Date().toISOString();
  const logEntry = { timestamp, step, message, level };

  let updateExpr =
    "SET #logs = list_append(if_not_exists(#logs, :empty_list), :new_log), " +
    "current_step = :step, #status = :status, updated_at = :ts";
  const exprValues = {
    ":new_log": [logEntry],
    ":empty_list": [],
    ":step": step,
    ":status": status,
    ":ts": timestamp,
  };
  const exprNames = {
    "#logs": "logs",
    "#status": "status",
  };

  if (result !== null) {
    updateExpr += ", #result = :result";
    exprValues[":result"] = result;
    exprNames["#result"] = "result";
  }

  await docClient.send(
    new UpdateCommand({
      TableName: process.env.PROGRESS_TABLE,
      Key: { application_id: applicationId },
      UpdateExpression: updateExpr,
      ExpressionAttributeValues: exprValues,
      ExpressionAttributeNames: exprNames,
    })
  );
}


// ─────────────────────────────────────────────────
// Utility Functions
// ─────────────────────────────────────────────────

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function getScenarioDecision(ssnLast4, loanAmount) {
  if (ssnLast4 === "1111") return "approved";
  if (ssnLast4 === "2222") return "denied";
  if (ssnLast4 === "3333") return loanAmount <= 25000 ? "approved" : "denied";
  return "approved";
}

function seededRandom(seed) {
  // Simple seeded PRNG using hash
  const hash = crypto.createHash("md5").update(seed).digest("hex");
  const value = parseInt(hash.substring(0, 8), 16);
  return value / 0xffffffff;
}

function seededRandInt(seed, min, max) {
  return Math.floor(seededRandom(seed) * (max - min + 1)) + min;
}


// ─────────────────────────────────────────────────
// Durable Steps
// ─────────────────────────────────────────────────

async function validateApplication(application) {
  logger.info("Validating application", { applicationId: application.application_id });
  await sleep(2000);

  const required = ["application_id", "applicant_name", "ssn_last4", "annual_income", "loan_amount", "loan_purpose"];
  const missing = required.filter((f) => !(f in application));
  if (missing.length > 0) {
    throw new Error(`Missing required fields: ${missing.join(", ")}`);
  }

  const income = application.annual_income;
  const loanAmount = application.loan_amount;

  if (loanAmount <= 0) throw new Error("Loan amount must be positive");
  if (income <= 0) throw new Error("Annual income must be positive");

  const dtiEstimate = (loanAmount * 0.05) / (income / 12);

  return {
    application_id: application.application_id,
    applicant_name: application.applicant_name,
    ssn_last4: application.ssn_last4,
    annual_income: income,
    loan_amount: loanAmount,
    loan_purpose: application.loan_purpose,
    estimated_dti: Math.round(dtiEstimate * 100) / 100,
    status: "validated",
    validated_at: new Date().toISOString(),
  };
}

async function pullCreditReport(bureau, ssnLast4) {
  logger.info(`Pulling credit report from ${bureau}`);
  await sleep(3000);

  const seed = `${bureau}-${ssnLast4}`;
  const score = seededRandInt(seed, 580, 820);
  const derogatoryMarks = seededRandInt(seed + "-derog", 0, 3);
  const openAccounts = seededRandInt(seed + "-accounts", 2, 15);

  return {
    bureau,
    score,
    report_id: `${bureau.substring(0, 3).toUpperCase()}-${ssnLast4}-${score}`,
    derogatory_marks: derogatoryMarks,
    open_accounts: openAccounts,
    pulled_at: new Date().toISOString(),
  };
}


async function calculateRiskScore(creditReports, ssnLast4, loanAmount) {
  logger.info("Calculating risk score from credit reports");
  await sleep(2000);

  const scores = creditReports.map((r) => r.score);
  const avgScore = scores.reduce((a, b) => a + b, 0) / scores.length;
  const totalDerogatory = creditReports.reduce((a, r) => a + r.derogatory_marks, 0);

  let tier, baseRate;
  if (avgScore >= 740 && totalDerogatory === 0) {
    tier = "prime";
    baseRate = 5.25;
  } else if (avgScore >= 670) {
    tier = "near-prime";
    baseRate = 7.5;
  } else if (avgScore >= 580) {
    tier = "subprime";
    baseRate = 11.0;
  } else {
    tier = "deep-subprime";
    baseRate = 15.0;
  }

  const decision = getScenarioDecision(ssnLast4, loanAmount);

  return {
    average_score: Math.round(avgScore * 10) / 10,
    min_score: Math.min(...scores),
    max_score: Math.max(...scores),
    total_derogatory_marks: totalDerogatory,
    risk_tier: tier,
    base_rate: baseRate,
    decision,
  };
}

async function generateLoanOffer(app, risk) {
  logger.info(`Generating offer for ${app.application_id}`);
  await sleep(2000);

  const rate = risk.base_rate;
  const loanAmount = app.loan_amount;
  const termMonths = 60;
  const monthlyRate = rate / 100 / 12;

  let payment;
  if (monthlyRate > 0) {
    payment =
      loanAmount *
      ((monthlyRate * Math.pow(1 + monthlyRate, termMonths)) /
        (Math.pow(1 + monthlyRate, termMonths) - 1));
  } else {
    payment = loanAmount / termMonths;
  }

  const offerHash = crypto
    .createHash("sha256")
    .update(`offer-${app.application_id}-${rate}`)
    .digest("hex")
    .substring(0, 10);

  return {
    offer_id: `OFFER-${offerHash.toUpperCase()}`,
    application_id: app.application_id,
    loan_amount: loanAmount,
    annual_rate: rate,
    term_months: termMonths,
    monthly_payment: Math.round(payment * 100) / 100,
    total_interest: Math.round((payment * termMonths - loanAmount) * 100) / 100,
    status: "offer_generated",
    generated_at: new Date().toISOString(),
  };
}


async function disburseFunds(offer) {
  logger.info(`Disbursing $${offer.loan_amount.toLocaleString()} for ${offer.offer_id}`);
  await sleep(2000);

  return {
    offer_id: offer.offer_id,
    disbursement_ref: `DSB-${offer.offer_id.slice(-6)}`,
    amount_disbursed: offer.loan_amount,
    status: "funded",
    funded_at: new Date().toISOString(),
  };
}


// ─────────────────────────────────────────────────
// Main Durable Execution Handler
// ─────────────────────────────────────────────────

export const handler = withDurableExecution(async (event, context) => {
  const docClient = getDocClient();
  const applicationId = event.application_id;
  logger.appendKeys({ applicationId });

  // Detect replay: count how many log entries exist per step in DynamoDB.
  const existing = await docClient.send(
    new GetCommand({
      TableName: process.env.PROGRESS_TABLE,
      Key: { application_id: applicationId },
    })
  );
  const priorLogs = (existing.Item && existing.Item.logs) || [];
  const priorCounts = {};
  for (const entry of priorLogs) {
    priorCounts[entry.step] = (priorCounts[entry.step] || 0) + 1;
  }
  const callCounts = {};

  async function log(step, message, status, level = "info", result = null) {
    callCounts[step] = (callCounts[step] || 0) + 1;
    if (callCounts[step] <= (priorCounts[step] || 0)) {
      message = `[REPLAY] ${message}`;
      level = "replay";
    }
    await logProgress(docClient, applicationId, step, message, status, level, result);
  }

  try {
    // ── Step 1: Validate Application ────────────────────
    await log("validating", "Validating loan application...", "processing");
    const validated = await context.step("validate-application", () => validateApplication(event));

    logger.info(
      `Application validated: ${validated.application_id} — $${validated.loan_amount.toLocaleString()} (${validated.loan_purpose})`
    );
    await log("validating", "Application validated successfully", "processing");


    // ── Step 2: Parallel Credit Bureau Checks ───────────
    await log("credit_check", "Pulling credit reports from 3 bureaus...", "processing");
    const bureaus = ["equifax", "transunion", "experian"];

    const creditReports = await context.parallel(
      bureaus.map((bureau) => ({
        name: `pull-credit-${bureau}`,
        fn: () => pullCreditReport(bureau, validated.ssn_last4),
      }))
    );

    const scoresStr = creditReports.map((r) => `${r.bureau}=${r.score}`).join(", ");
    logger.info(`Credit reports pulled — scores: ${scoresStr}`);
    await log("credit_check", `Credit scores received: ${scoresStr}`, "processing");

    // ── Step 3: Risk Assessment ─────────────────────────
    await log("risk_assessment", "Calculating risk score...", "processing");
    const risk = await context.step("calculate-risk-score", () =>
      calculateRiskScore(creditReports, validated.ssn_last4, validated.loan_amount)
    );

    logger.info(`Risk assessed: tier=${risk.risk_tier}, avg=${risk.average_score}, decision=${risk.decision}`);
    await log(
      "risk_assessment",
      `Risk tier: ${risk.risk_tier}, avg score: ${risk.average_score}, decision: ${risk.decision}`,
      "processing"
    );

    // ── Denied Path ─────────────────────────────────────
    if (risk.decision === "denied") {
      const finalResult = {
        application_id: validated.application_id,
        applicant_name: validated.applicant_name,
        status: "denied",
        reason: `Application denied — risk tier: ${risk.risk_tier}, avg credit score: ${risk.average_score}`,
        risk_tier: risk.risk_tier,
        average_score: risk.average_score,
      };
      await log("risk_assessment", "Application denied", "denied", "warn", finalResult);
      return finalResult;
    }


    // ── Step 4: Manager Approval (if >= $100,000) ────────
    if (validated.loan_amount >= 100000) {
      await log(
        "manager_approval",
        `Manager approval required for loans >= $100,000 (requested: $${validated.loan_amount.toLocaleString()})`,
        "pending_approval"
      );

      const approvalResult = await context.waitForCallback(
        "manager-approval",
        async (callbackId) => {
          // Store callback_id in DynamoDB so the frontend can send the approval
          await docClient.send(
            new UpdateCommand({
              TableName: process.env.PROGRESS_TABLE,
              Key: { application_id: validated.application_id },
              UpdateExpression: "SET callback_id = :cid",
              ExpressionAttributeValues: { ":cid": callbackId },
            })
          );
        },
        { timeout: 30 * 60 * 1000 } // 30 minutes
      );

      const approval = typeof approvalResult === "string" ? JSON.parse(approvalResult) : approvalResult;
      logger.info("Manager approval result", { approval });

      if (!approval.approved) {
        const finalResult = {
          application_id: validated.application_id,
          applicant_name: validated.applicant_name,
          status: "denied",
          reason: approval.reason || "Manager denied the application",
        };
        await log("manager_approval", "Application denied by manager", "denied", "warn", finalResult);
        return finalResult;
      }

      await log("manager_approval", "Manager approved the application", "processing");
    }


    // ── Step 5: External Fraud Check (Callback) ─────────
    // The workflow SUSPENDS here. An external Lambda (FraudCheckFunction)
    // processes the request and calls SendDurableExecutionCallbackSuccess
    // to resume this execution. Zero compute cost while waiting.
    await log("fraud_check", "Requesting external fraud check service...", "processing");

    const fraudResult = await context.waitForCallback(
      "fraud-check",
      async (callbackId) => {
        // Invoke the external fraud check Lambda, passing the callback_id
        const lambdaClient = new LambdaClient({});
        await lambdaClient.send(
          new InvokeCommand({
            FunctionName: process.env.FRAUD_CHECK_FUNCTION,
            InvocationType: "Event",
            Payload: Buffer.from(
              JSON.stringify({
                callback_id: callbackId,
                application_id: validated.application_id,
                applicant_name: validated.applicant_name,
              })
            ),
          })
        );
      },
      { timeout: 5 * 60 * 1000 } // 5 minutes
    );

    const fraud = typeof fraudResult === "string" ? JSON.parse(fraudResult) : fraudResult;
    logger.info("Fraud check result", { fraud });
    await log(
      "fraud_check",
      `Fraud check passed — ${fraud.checked_by || "external service"}`,
      "processing"
    );


    // ── Step 6: Generate Loan Offer ─────────────────────
    await log("generating_offer", "Generating loan offer...", "processing");
    const offer = await context.step("generate-loan-offer", () => generateLoanOffer(validated, risk));

    logger.info(`Offer generated: ${offer.offer_id} — $${offer.monthly_payment}/mo at ${offer.annual_rate}%`);
    await log(
      "generating_offer",
      `Offer ${offer.offer_id}: $${offer.monthly_payment}/mo at ${offer.annual_rate}%`,
      "processing"
    );

    // ── Step 7: Disburse Funds ──────────────────────────
    await log("disbursing", "Disbursing funds...", "processing");
    const disbursement = await context.step("disburse-funds", () => disburseFunds(offer));

    logger.info(`Funds disbursed: ${disbursement.disbursement_ref}`);

    const finalResult = {
      application_id: validated.application_id,
      applicant_name: validated.applicant_name,
      status: "approved",
      offer_id: offer.offer_id,
      loan_amount: offer.loan_amount,
      annual_rate: offer.annual_rate,
      monthly_payment: offer.monthly_payment,
      term_months: offer.term_months,
      disbursement_ref: disbursement.disbursement_ref,
    };
    await log("complete", "Loan approved and funds disbursed!", "approved", "info", finalResult);
    return finalResult;
  } catch (error) {
    logger.error("Loan workflow failed", { error: error.message });
    await log("error", `Workflow error: ${error.message}`, "failed", "error");
    throw error;
  }
});
